### PR TITLE
Route deploy from speculative-exec server to `DeployAcceptor`

### LINKED ISSUE #3969: Route speculative exec deploys to DeployAcceptor
deploys arriving via the speculative exec server should initially be routed to the DeployAcceptor for normal validation and verification.  The Source enum should be given a new variant to represent this source, and in the acceptor, this will be handled as per Source::Client up until the point of cryptographic verification.  If this succeeds, the responder will be called and no further action by the acceptor is required (i.e. the deploy is not stored, and no announcement is made).  Once the responder is called, the speculative exec server will resume the existing flow by passing the deploy to contract runtime for execution.

### PR DESCRIPTION
**Pull request by Fraser999**
This PR primarily updates the speculative-exec server to pass incoming deploys to the `DeployAcceptor` before attempting speculative execution.  This more closely matches the normal flow for incoming deploys, and means such deploys will undergo more rigorous validation, including cryptographic verification of the approvals.

This required a little reworking of `DeployAcceptor` to ensure the speculative-exec deploys are not stored on the node or announced to other components.  It also meant that the `DeployAcceptor` had to be able to use a specific block for validation, rather than just the highest complete block, as the speculative-exec endpoint allows clients to specify a block at which to attempt execution.

There is also a fix for nctl, so that all nodes by default enable their speculative-exec servers, on ports 25101, 25102, etc.

Closes #3969.

### CODE REVIEW

File: utils/nctl/sh/assets/setup.sh [hunk 1]
```diff
@@ -86,16 +84,8 @@ function _set_nodes()
...
-        )
-
-        if [ ! -z "$SPECULATIVE_EXEC_ADDR" ]; then
-            SCRIPT+=(
-                "cfg['rpc_server']['speculative_execution_address']='0.0.0.0:$(get_node_port_speculative_exec "$IDX")';"
-            )
-        fi
-
-        SCRIPT+=(
+            "cfg['speculative_exec_server']['address']='0.0.0.0:$(get_node_port_speculative_exec "$IDX")';"
```
Comment (rafal-ch) [inline utils/nctl/sh/assets/setup.sh]: I know it's not a scope of this PR, but can you consider extending the `nctl-view-node-ports` to also show the speculative exec port number?

File: node/src/components/deploy_acceptor.rs [hunk 2]
```diff
@@ -272,10 +273,34 @@ impl DeployAcceptor {
             }
         }
 
+        // If this has been received from the speculative exec server, use the block specified in
+        // the request, otherwise use the highest complete block.
```
Comment (rafal-ch) [inline node/src/components/deploy_acceptor.rs:274]: In the **non-speculative** flow, if the block id is missing we use the `get_highest_complete_block_header_from_storage()`, which uses the `Storage::completed_blocks` to find the most recent completed block.

In the speculative flow, when we don't provide a block, we'll use the `get_highest_block_with_metadata_from_storage()`, which uses the `Storage::block_height_index` to get the most recent block.

I was wondering if the latter should also use the most recent **complete** block?

**Review by rafal-ch — COMMENTED** [review]
Not a full review yet, but can't continue at the moment.

File: utils/nctl/sh/assets/setup.sh (hunk 1 above)
Comment (Fraser999) [reply utils/nctl/sh/assets/setup.sh]: Done in [e75c9c8](https://github.com/casper-network/casper-node/pull/3972/commits/e75c9c8098f48970f4cd0722d670a3799f60f995).

File: node/src/components/deploy_acceptor.rs (hunk 2 above)
Comment (Fraser999) [reply node/src/components/deploy_acceptor.rs:274]: Good catch!

Fixed in [3eb3d54](https://github.com/casper-network/casper-node/pull/3972/commits/3eb3d54ad72c10fdc7f47479d78683966d95a2fc).

File: node/src/types/deploy/deploy_header.rs [hunk 3]
```diff
@@ -91,13 +92,49 @@ impl DeployHeader {
...
+                max_ttl = %config.max_ttl,
+                "deploy ttl excessive"
+            );
+            return Err(DeployConfigurationFailure::ExcessiveTimeToLive {
+                max_ttl: config.max_ttl,
+                got: self.ttl(),
+            });
+        }
+
+        if self.timestamp() > at {
```
Comment (rafal-ch) [inline node/src/types/deploy/deploy_header.rs:129]: Shall we consider allowing a potential, small gap to compensate for the fact that clocks are never perfectly in sync? So I'd say it should be ok if the deploy timestamp is, let's say, 5 sec. in the future?

I think a similar gap is allowed in authenticator kind of apps.

Comment (alsrdn) [reply node/src/types/deploy/deploy_header.rs:129]: I guess the node's clock could be a bit behind the client. Though here `at` will be higher than `self.timestamp()` if the clocks are perfectly in sync because of the network delay for sending the deploy. But since we strive to have the clocks in sync, the difference would be on the order of tens or hundreds of miliseconds at most.
I would say we shouldn't add additional complexity here. Since the client wants to communicate with this node, it can account for the clock difference itself and delay the creation of the deploy by a few miliseconds if it receives feedback that the deploy is in the future.

File: node/src/components/deploy_acceptor.rs [hunk 4]
```diff
@@ -242,6 +242,7 @@ impl DeployAcceptor {
             &self.chain_name,
             &self.deploy_config,
             self.max_associated_keys,
+            verification_start_timestamp,
```
Comment (alsrdn) [inline node/src/components/deploy_acceptor.rs:245]: Not part of this PR but we probably want to use `verification_start_timestamp` when `source.is_client()` also since the error for expired deploy uses it too.

File: node/src/types/deploy/deploy_header.rs (hunk 3 above)
Comment (Fraser999) [reply node/src/types/deploy/deploy_header.rs:129]: Yup, I agree with @alsrdn.

File: node/src/components/deploy_acceptor.rs [hunk 5]
```diff
@@ -856,6 +881,18 @@ impl DeployAcceptor {
             );
         }
 
+        // If this has been received from the speculative exec server, we just want to call the
+        // responder and finish.  Otherwise store the deploy and announce it if required.
+        if let Source::SpeculativeExec(_) = event_metadata.source {
+            let effects = if let Some(responder) = event_metadata.maybe_responder {
```
Comment (alsrdn) [inline node/src/components/deploy_acceptor.rs:884]: Should we also increment the `accepted` metric here as well since it is only incremented when we put the deploy to storage? We do increment the `rejected` count if the deploy is invalid. Or maybe we want to have a separate metric for speculative deploys?

**Review by alsrdn — COMMENTED** [review]
Just one question related to metrics, otherwise LGTM.

File: node/src/components/deploy_acceptor.rs (hunk 5 above)
Comment (Fraser999) [reply node/src/components/deploy_acceptor.rs:884]: Hmm - good question.

I'd be inclined to not increment either existing metric for speculative execution, and to not provide any metrics around that flow until we see a clear benefit to having them.  I'll wait for others to weight in, but I definitely dislike the current asymmetry you noticed.

File: node/src/components/deploy_acceptor.rs (hunk 4 above)
Comment (Fraser999) [reply node/src/components/deploy_acceptor.rs:245]: You mean instead of `current_node_timestamp`?  I agree, and I'd be inclined to make that change in this PR too.

@rafal-ch - any objections if I do that?

File: node/src/components/deploy_acceptor.rs (hunk 5 above)
Comment (fizyk20) [reply node/src/components/deploy_acceptor.rs:884]: I'm for not incrementing either one for speculative execution :+1:

File: node/src/components/deploy_acceptor.rs (hunk 4 above)
Comment (Fraser999) [reply node/src/components/deploy_acceptor.rs:245]: Done in [b25ef5b](https://github.com/casper-network/casper-node/pull/3972/commits/b25ef5b0a542ed178b234bcefdb771fba9c8efb2).

File: node/src/components/deploy_acceptor.rs (hunk 5 above)
Comment (Fraser999) [reply node/src/components/deploy_acceptor.rs:884]: Done in [b25ef5b](https://github.com/casper-network/casper-node/pull/3972/commits/b25ef5b0a542ed178b234bcefdb771fba9c8efb2).

**Review by alsrdn — APPROVED** [review]

**Comment by Fraser999** [comment]
bors merge
